tempesta_fw/t/unit/test_sched_hash.c

```diff
@@ -142,64 +136,50 @@ TEST(tfw_sched_hash, one_srv_in_sg_and_max_conn)
 
 TEST(tfw_sched_hash, max_srv_in_sg_and_zero_conn)
 {
-	int i, j;
-
-	TfwSrvGroup *sg = test_create_sg("test", "hash");
-
-	for (i = 0; i < TFW_SG_MAX_SRV; ++i) {
-		test_create_srv("127.0.0.1", sg);
-	}
-
-	for (i = 0; i < req_strs_size; ++i) {
-		for (j = 0; j < 2 * TFW_SG_MAX_SRV; ++j) {
-			TfwConnection *conn;
-			TfwHttpReq *req = test_req_alloc(strlen(req_strs[i]));
-
-			tfw_http_parse_req(req, req_strs[i], strlen(req_strs[i]));
-
-			conn = sg->sched->sched_srv((TfwMsg *)req, sg);
-			EXPECT_TRUE(conn == NULL);
-
-			test_req_free(req);
-		}
-	}
-
-	test_sg_release_all();
+	test_sched_generic_max_srv_in_sg_and_zero_conn(&sched_helper_hash);
 }
 
+/*
+ * This unit test is implementation aware and checks more than just interface.
+ * Note, that it is very similar to other tests (one_srv_in_sg_and_max_conn and
+ * max_srv_in_sg_and_max_conn) for round-robin and hash schedullers. So if test
+ * structure is changed, other mentioned in above tests should be also be updated
+ */
 TEST(tfw_sched_hash, max_srv_in_sg_and_max_conn)
 {
-	int i, j;
-	TfwSrvGroup *sg = test_create_sg("test", "hash");
-	TfwSrvConnection *sconn;
+	size_t i, j;
+
+	TfwSrvGroup *sg = test_create_sg("test", sched_helper_hash.sched);
 
 	for (i = 0; i < TFW_SG_MAX_SRV; ++i) {
 		TfwServer *srv = test_create_srv("127.0.0.1", sg);
 
 		for (j = 0; j < TFW_SRV_MAX_CONN; ++j) {
-			sconn = test_create_conn((TfwPeer *)srv);
+			TfwSrvConnection *sconn = test_create_conn((TfwPeer *)srv);
+
 			sg->sched->add_conn(sg, srv, &sconn->conn);
+			/* A connection is skipped by schedulers if (refcnt <= 0). */
+			tfw_connection_get(&sconn->conn);
 		}
 	}
 
-	for (i = 0; i < req_strs_size; ++i) {
-		TfwConnection *s = NULL;
+	/* Check that every request is scheduled to the same connection. */
+	for (i = 0; i < sched_helper_hash.conn_types; ++i) {
+		TfwConnection *expected_connection = NULL;
```

Comment:
Not a problem, but usually we don't use long names for functions and variables, for example `exp_conn`'d be better.